psh/dd/dd.c

```diff
@@ -0,0 +1,342 @@
+/*
+ * Phoenix-RTOS
+ *
+ * dd - converts and copies a file according to operands
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <ctype.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <limits.h>
+
+#include "../psh.h"
+
+/* clang-format off */
+
+enum { param_none = 0, param_if, param_of, param_bs,
+	param_count, param_seek, param_skip };
+
+
+struct param_s {
+	char *name;
+	int value;
+};
+
+
+static const struct param_s params[] = {
+	{ "if", param_if },
+	{ "of", param_of },
+	{ "bs", param_bs },
+	{ "count", param_count },
+	{ "seek", param_seek },
+	{ "skip", param_skip },
+	{ NULL, param_none }
+};
+
+/* clang-format on */
+
+
+static void psh_dd_info(void)
+{
+	printf("copy a file according to the operands");
+}
+
+
+static void psh_dd_usage(void)
+{
+	printf(
+		"Usage: dd [OPERAND]...\n"
+		"\tif=FILE   read from FILE instead of stdin\n"
+		"\tof=FILE   write to FILE instead of stdout\n"
+		"\tbs=BYTE   read/write block size of BYTES bytes\n"
+		"\tcount=N   copy only N input blocks\n"
+		"\tseek=N    skip N bs-sized blocks at start of output\n"
+		"\tskip=N    skip N bs-sized blocks at start of input\n");
+}
+
+
+static ssize_t getnumber(char *numstr)
+{
+	ssize_t value;
+
+	if (isdigit(*numstr) == 0) {
+		return -EINVAL;
+	}
+
+	value = 0;
+	while (isdigit(*numstr) != 0) {
+		value = value * 10 + *(numstr++) - '0';
+	}
+
+	switch (*(numstr++)) {
+		case 'M':
+			value *= 1024 * 1024;
+			break;
+
+		case 'k':
+			value *= 1024;
+			break;
+
+		case 'b':
+			value *= 512;
+			break;
+
+		case 'w':
+			value *= 2;
+			break;
+
+		case 'c':
+			/* value *= 1; */
+			break;
+
+		case '\0':
+			return value;
+
+		default:
+			return -1;
+	}
+
+	if (*numstr != '\0') {
+		return -EINVAL;
+	}
+
+	return value;
+}
+
+
+static int psh_dd(int argc, char **argv)
+{
+	const struct param_s *par;
+	char *buf, *str, *cp;
+	int infd, outfd, incc, outcc;
+	ssize_t intotal, outtotal;
+
+	char *infile = NULL;
+	char *outfile = NULL;
+	ssize_t count = SSIZE_MAX;
+	ssize_t seekval = 0;
+	ssize_t skipval = 0;
+	ssize_t inmax = 0;
+	int err = 0, blocksz = 512;
+
+	if ((argc > 1) && (argv[1][0] == '-') && (argv[1][1] == 'h')) {
+		psh_dd_usage();
+		return EXIT_SUCCESS;
+	}
+
+	while (--argc > 0) {
+		str = *(++argv);
+		cp = strchr(str, '=');
+		if (cp == NULL) {
```

Comment:
May add this to check an empty string after the `=` char.
```suggestion
		if ((cp == NULL) || (cp[1] == '\0')){
```